File: epple2qt/src/slots.cpp
/***************************************************************************
 *   Copyright (C) 2008 by Chris Mosher,,,   *
 *   chris@mosher.mine.nu   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include "slots.h"

const int Slots::SLOTS(8);

Slots::Slots():
	cards(SLOTS,&this->empty)
{
}

Slots::~Slots()
{
}

unsigned char Slots::io(const int islot, const int iswch, const unsigned char b, const bool writing)
{
	return this->cards[islot]->io(iswch,b,writing);
}

struct resetCard
{
	void operator() (Card* p) { p->reset(); }
};

void Slots::reset()
{
	std::for_each(this->cards.begin(),this->cards.end(),resetCard());
}

unsigned char Slots::readRom(const int islot, const unsigned short addr)
{
	return this->cards[islot]->readRom(addr);
}

struct readSeventhRomCard
{
	const unsigned short addr;
	unsigned char* b;
	readSeventhRomCard(const unsigned short addr, unsigned char* b):addr(addr),b(b){}
	void operator() (Card* p) { p->readSeventhRom(this->addr,this->b); }
};

unsigned char Slots::readSeventhRom(const unsigned short addr)
{
	unsigned char b(0xFF);
	std::for_each(this->cards.begin(),this->cards.end(),readSeventhRomCard(addr,&b));
	return b;
}

struct ioBankRomCard
{
	const unsigned short addr;
	unsigned char* b;
	const bool write;
	ioBankRomCard(const unsigned short addr, unsigned char* b, const bool write):addr(addr),b(b),write(write){}
	void operator() (Card* p) { p->ioBankRom(this->addr,this->b,this->write); }
};

unsigned char Slots::ioBankRom(const unsigned short addr, const unsigned char data, const bool write)
{
	unsigned char b(data);
	std::for_each(this->cards.begin(),this->cards.end(),ioBankRomCard(addr,&b,write));
	return b;
}

struct inhibitMotherboardRomCard
{
	bool inhibit;
	inhibitMotherboardRomCard():inhibit(false) {}
	void operator() (Card* p) { if (p->inhibitMotherboardRom()) inhibit = true; }
};

bool Slots::inhibitMotherboardRom()
{
	inhibitMotherboardRomCard inh = inhibitMotherboardRomCard();
	std::for_each(this->cards.begin(),this->cards.end(),inh);
	return inh.inhibit;
}

void Slots::set(const int slot, Card* card)
{
	this->cards[slot] = card;
}

Card* Slots::get(const int slot)
{
	return this->cards[slot];
}


/*
struct isAnyDiskDriveMotorOnCard
{
	bool on;
	isAnyDiskDriveMotorOnCard():on(false) {}
	void operator() (Card* p) { if (p->isMotorOn()) on = true; }
};

bool isAnyDiskDriveMotorOn()
{
	isAnyDiskDriveMotorOnCard on = isAnyDiskDriveMotorOnCard();
	std::for_each(this->cards.begin(),this->cards.end(),inh);
	return on.inhibit;
}
*/
struct isDirtyCard
{
	bool dirty;
	isDirtyCard():dirty(false) {}
	void operator() (Card* p) { if (p->isDirty()) dirty = true; }
};

bool Slots::isDirty()
{
	isDirtyCard dirty = isDirtyCard();
	std::for_each(this->cards.begin(),this->cards.end(),dirty);
	return dirty.dirty;
}

File: apple2/emu/epple2/old_java_version/ja2/src/chipset/Slots.java
/*
 * Created on Nov 30, 2007
 */
package chipset;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import cards.Card;
import cards.EmptySlot;
import cards.disk.DiskController;

public class Slots implements Iterable<Card>
{
	public static final int SLOTS = 8;

	private final List<Card> cards = new ArrayList<Card>(SLOTS);


	public Slots()
	{
		this.cards.addAll(Collections.<Card>nCopies(SLOTS,new EmptySlot()));
	}

	public byte io(final int islot, final int iswch, final byte b, boolean writing)
	{
		return this.cards.get(islot).io(iswch,b,writing);
	}

	public void reset()
	{
		for (final Card card: this.cards)
		{
			card.reset();
		}
	}

	public byte readRom(final int islot, final int addr)
	{
		return this.cards.get(islot).readRom(addr);
	}

	private final byte[] rb = new byte[1];
	public byte readSeventhRom(int addr)
	{
		this.rb[0] = -1;

		for (final Card card : this.cards)
		{
			card.readSeventhRom(addr,this.rb);
		}

		return this.rb[0];
	}

	public byte ioBankRom(final int addr, byte data, boolean write)
	{
		this.rb[0] = data;

		for (final Card card : this.cards)
		{
			card.ioBankRom(addr,this.rb,write);
		}

		return this.rb[0];
	}

	public boolean inhibitMotherboardRom()
	{
		for (final Card card: this.cards)
		{
			if (card.inhibitMotherboardRom())
			{
				return true;
			}
		}
		return false;
	}

	public void set(final int slot, final Card card)
	{
		this.cards.set(slot,card);
	}

	public Card get(final int slot)
	{
		return this.cards.get(slot);
	}

	public Iterator<Card> iterator()
	{
		return this.cards.iterator();
	}

	public boolean isAnyDiskDriveMotorOn()
	{
		for (final Card card: this.cards)
		{
			if (card instanceof DiskController)
			{
				final DiskController disk = (DiskController)card;
				if (disk.isMotorOn())
				{
					return true;
				}
			}
		}
		return false;
	}

	public boolean isAnyDiskDirty()
	{
		for (final Card card: this.cards)
		{
			if (card instanceof DiskController)
			{
				final DiskController disk = (DiskController)card;
				if (disk.isDirty())
				{
					return true;
				}
			}
		}
		return false;
	}
}
